dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,670 @@
-// directly contain a either a DeviceClaimName (to enable multiple pods to
-// refer to a pre-provisioned claim), or an embedded struct that includes
-// ObjectMeta and a DeviceClaimName. In this case, the named DeviceClaim will
-// be treated as a template; that is, its spec will be copied to create a new
-// claim, based on the new metadata. Re-using claim-as-a-template avoids
-// another, nearly identical top-level API object. But it may be confusing, we
-// need feedback.
-//
-// DeviceClassConfigReference is used to refer to arbitrary configuration
-// objects from the class. Since it is the class, and therefore is created by
-// the administrator, it allows referencing objects in any namespace.
+	// NodeName is the name of the node providing the necessary resources
+	// if the resources are local to a node.
+	//
+	// +optional
+	NodeName string `json:"nodeName,omitempty" protobuf:"bytes,3,name=nodeName"`
 
-type DeviceClassConfigReference struct {
-	// API version of the referent.
-	// +required
-	APIVersion string `json:"apiVersion"`
+	// Results lists all allocated driver resources.
+	//
+	// +listType=atomic
+	Results []RequestAllocationResult `json:"results" protobuf:"bytes,4,name=results"`
+}
 
-	// Kind of the referent.
-	// +required
-	Kind string `json:"kind"`
+// DriverConfiguration is one entry in a list of configuration pieces.
+type DriverConfiguration struct {
+	// Admins is true if the source of the piece was a class and thus
+	// not something that a normal user would have been able to set.
+	Admin bool `json:"admin,omnitempty"`
 
-	// Namespace of the referent.
-	// +required
-	Namespace string `json:"namespace"`
+	DriverConfigurationAlternatives `json:",inline"`
+}
 
-	// Name of the referent.
-	// +required
-	Name string `json:"name"`
+// DriverConfigurationAlternatives must have one and only one one field set.
+//
+// In contrast to VendorConfigurationParameters, the driver name is
+// not included and has to be infered from the context.
+type DriverConfigurationAlternatives struct {
+	Vendor *runtime.RawExtension `json:"vendor,omitempty" protobuf:"bytes,1,opt,name=vendor"`
 }
 
-// DeviceConfigReference is used to refer to arbitrary configuration object
-// from the claim. Since it is created by the end user, the referenced objects
-// are restricted to the same namespace as the DeviceClaim.
-type DeviceConfigReference struct {
-	// API version of the referent.
-	// +required
-	APIVersion string `json:"apiVersion"`
+// RequestAllocationResult contains configuration and the allocation result for
+// one request.
+type RequestAllocationResult struct {
+	// Config contains all the configuration pieces that apply to the request
+	// and that were meant for the driver which handles these resources.
+	// They get collected during the allocation and stored here
+	// to ensure that they remain available while the claim is allocated.
+	//
+	// Entries are sorted from "most specific" first to "least specific" last:
```

Comment:
Why not separate fields? That way the driver can decide which takes precedence.

Reply:
Wouldn't it be better to standardize some common expectation for precendence instead of leaving it to the driver authors?

Regarding separate fields: I had that earlier but then dropped it when adding multi-inheritance. I did not revert back because a) a single field seemed easier b) it allows us to change the UX without affecting drivers.

Reply:
With separate fields, the driver author can be confident that the "class" related items are owned by the *administrator*. So, they can take that into account as to what the *user* is allowed to do vs the *admin*.

Reply:
That is what `DriverConfiguration` has `Admin bool` for.

Reply:
Ok, maybe I missed that. So, there is no way for the user to populate something in here that looks like it was from the admin?

Reply:
Correct.